[base][docs] Base UI Quickstart Guide

mj12albert opened the pull request:
Closes https://github.com/mui/material-ui/issues/32979

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

This PR addresses all 3 points from https://github.com/mui/material-ui/issues/32979 by expanding the "Quickstart" to demo how to style a button with a few styling solutions (plain CSS, Tailwind, MUI system) using mostly snippets (not complete code), though complete demos are still provided for each.

Additionally, I made a separate Quickstart page that includes the installation instructions too. Base is one package and the installation instructions will become even shorter after [this](https://github.com/mui/material-ui/pull/36766), it seems a bit much to have a page just for that (despite the fact that our other docs have an Installation page, but they are all more complex).

mj12albert commented on docs/src/modules/components/MarkdownDocs.js:
```diff
@@ -103,6 +103,18 @@ export default function MarkdownDocs(props) {
 
           const name = renderedMarkdownOrDemo.demo;
+
+          if (renderedMarkdownOrDemo.excludeByCurrentColorMode) {
+            const matchedColorMode = name.match(/([^/.]+(?=\.js|jsx|ts|tsx$))/)[0];
+
+            if (
+              ['light', 'dark'].includes(matchedColorMode) &&
+              matchedColorMode !== theme.palette.mode
+            ) {
+              return <span key={`${index}-excluded-by-current-color-mode`} />;
+            }
+          }
```
I would like to only show `Demo.light.js` and exclude `Demo.dark.js` if the docs is currently in light mode (and vice versa)

Can anybody think of a less hacky way or is this more or less OK 😅 ? @mui/core

mj12albert replied:
I don't need to do this anymore

siriwatknp reviewed (commented):
@mj12albert For the light & dark, I suggest using CSS variables instead of having 2 files with javascript theme.

It could look something like this:

```js
const getStyles = () => ({
  --btn-background: #1f883d,
  '.mode-dark': { // class name `mode-dark` is handled by the docs
    --btn-background: #238636,
  },
  background-color: 'var(--btn-background)',
})

const GithubButtonComponent = styled(ButtonUnstyled)(getStyles);
```

mj12albert commented:
Demo: https://deploy-preview-36717--material-ui.netlify.app/base/getting-started/quickstart/

I'm still polishing the demos and snippets, but the content itself is ready for an early review - please have a look and welcome feedback on anything that catches your eye! @michaldudak @mnajdova @samuelsycamore

mj12albert commented:
> @mj12albert For the light & dark, I suggest using CSS variables instead of having 2 files with javascript theme.

Let me try this – I wanted to separate them completely at first because I didn't want the snippets to show anything related to color mode at this point. But since I'm mostly not using the generated one anyway it probably doesn't matter in the end.

danilo-leal commented:
This is looking excellent, Albert ⎯ excited to see it shaping up! To start off, one high-level thing I thought it'd be better is separating the Tutorial from the Installation page, having it be a standalone page as it is with Joy. What do you think?

mj12albert commented:
> This is looking excellent, Albert ⎯ excited to see it shaping up! To start off, one high-level thing I thought it'd be better is separating the Tutorial from the Installation page, having it be a standalone page as it is with Joy. What do you think?

🤔 Perhaps I should not use `Tutorial` as a heading here, I do plan to have a standalone tutorial page thats more or less equivalent to Joy, but with something more complex than a simple button ~

mj12albert commented:
@danilo-leal @samuelsycamore I changed the heading to not use the word "tutorial" and updated the PR description to be clearer 😅

This is meant to be an expansion of the [current](https://mui.com/base/getting-started/usage/#quickstart) quickstart section, and not a step-by-step tutorial like the Joy UI tutorial.

So it totally makes sense to remain on the "Usage" page.

But I wonder if Base really needs a separate installation page (just because the other packages have one). It's only one package and the page will become even shorter after https://github.com/mui/material-ui/pull/36766

PS I have done the style guide edits

danilo-leal commented:
> But I wonder if Base really needs a separate installation page (just because the other packages have one). It's only one package and the page will become even shorter after https://github.com/mui/material-ui/pull/36766

Yeah, it definitely makes me think that, given the new installation code block, the "Installation" + "Usage" page across all docs could be unified. I'd be down for this approach! It's important we keep consistency, though, regardless of which we end up picking!

danilo-leal commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,244 @@
+
+## Implementing a Button
+
+This is a quick tutorial that goes through the basics of using and styling Base components by replicating a button from Github's UI, using their [style guide](https://primer.style) as a reference.
```
```suggestion
This is a quick tutorial that goes through the basics of using and styling Base components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.
```

Sorry ⎯ such a pedantic change but it's important to me 😅 😆

michaldudak replied:
```suggestion
This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from Github's UI, using their [style guide](https://primer.style) as a reference.
```

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
````diff
@@ -0,0 +1,244 @@
+}
+```
+
+#### Styling useButton with MUI System
````
```suggestion
#### useButton hook with MUI System
```

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
````diff
@@ -0,0 +1,244 @@
+}
+```
+
+#### Using the sx prop
````
I think this is worth upgrading to an h3 so it sits at the same level as the other styling options.

```suggestion
### Styling with the sx prop
```

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
````diff
@@ -0,0 +1,244 @@
+}
+```
+
+Here's a complete demo of styling ButtonUnstyled and useButton with MUI System:
````
```suggestion
The demo below shows how to build the Primer button using MUI System along with the `sx` prop:
```

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,244 @@
+
+{{"demo": "BaseButtonMuiSystem.js", "defaultCodeOpen": false}}
+
+Read more about working with MUI System and its utilities [here](/system/getting-started/usage/).
```
```suggestion
Read the [MUI System Usage](/system/getting-started/usage/) doc for further details.
```

samuelsycamore reviewed (commented):
This review is mostly just applying the style guide—the content is looking great!

michaldudak commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,244 @@
+  const buttonRef = React.useRef();
+
+  const { getRootProps } = useButton({
+    ref: buttonRef,
```
There's no need to pass a ref if you don't do anything with it. `useButton({})` should be enough in this case (but I just realized it looks weird and we should allow `useButton()`)

+ other places that show to use a hook

mj12albert replied:
👌 Got it, I've opened some related issues:
- https://github.com/mui/material-ui/issues/36880
- https://github.com/mui/material-ui/issues/36879

siriwatknp commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,244 @@
+
+[`react`](https://www.npmjs.com/package/react) >= 17.0.0 and [`react-dom`](https://www.npmjs.com/package/react-dom) >= 17.0.0 are peer dependencies.
+
+## Implementing a Button
```
I wonder if it is worth adding a paragraph that explains why you should use Base UI button instead of a plain HTML button.

I guess a lot of people might have this question in their minds. cc @samuelsycamore

samuelsycamore replied:
Great idea @siriwatknp !

mj12albert commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,245 @@
+## Implementing a Button
+
+This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.
+Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling etc while remaining unopinionated about styling, giving you complete control over your app's CSS.
```
> I wonder if it is worth adding a paragraph that explains why you should use Base UI button instead of a plain HTML button.

@samuelsycamore I've added a sentence here, not sure if it's the smoothest transition 😅

danilo-leal replied:
Should we add a heading to separate these two? Quite directly "Why use Base UI's button instead of native HTML?" or something like that. 😬

samuelsycamore replied:
@danilo-leal I think that's a great idea, but I wonder if this is the best place for it? I would assume that if you've made it to the Quickstart, you've already concluded that you want to use Base's Button because it's better than an HTML button. 😁 Maybe this calls for an update to the Overview page?

samuelsycamore replied:
Avoid Latinisms whenever possible 😇

```suggestion
Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling, and more, while remaining unopinionated about styling, giving you complete control over your app's CSS.
```

danilo-leal replied:
Oh, yeah ⎯ that checks out! Might make sense to tackle it together with the "pro vs. con" thread that was going on. Base UI should add some advantage against native HTML 😬

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,245 @@
+### Components and hooks
+
+Base UI provides an Unstyled Button component and a `useButton` hook.
+Both can be used to build a button, and each has its own benefits and trade-offs — see [Components vs. hooks](/base/getting-started/usage/#components-vs-hooks) for details.
```
No spaces on either side of an em dash (—).

```suggestion
Both can be used to build a button, and each has its own benefits and trade-offs—see [Components vs. hooks](/base/getting-started/usage/#components-vs-hooks) for details.
```

samuelsycamore commented on docs/data/base/getting-started/quickstart/quickstart.md:
````diff
@@ -0,0 +1,245 @@
+}
+```
+
+Base UI comes with no styles or styling solution - here's what the Unstyled Button looks like out of the box:
````
Em dash.

```suggestion
Base UI comes with no styles or styling solution—here's what the Unstyled Button looks like out of the box:
```

samuelsycamore reviewed (commented):
Almost there! 🙌

mj12albert commented:
@mui/core bumping this for a re-review, I think it should be good to go ~ 😬

danilo-leal reviewed (approved):
Looks great overall! Awesome job 🤙 
Left just a tiny comment there.

michaldudak commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,235 @@
+
+### Components and hooks
+
+Base UI provides a Button component and a `useButton` hook.
```
```suggestion
Base UI provides a Button component and a useButton hook.
```

or "a `Button` component and a `useButton` hook"

samuelsycamore replied:
@michaldudak according to the [Style Guide](https://www.notion.so/mui-org/Writing-style-guide-2a957a4168a54d47b14bae026d06a24b), the way it's written currently is correct. (Of course it can always be changed as we see fit.)

If you wanted to put both in backticks, the preferred style would be `<Button />` and `useButton`. If you think that neither of them should have backticks, then there are many other instances of `useSomeHook` throughout the Base and Joy docs that we'll want to update to remove backticks.

mj12albert replied:
Let's go with `<Button />` here as it's the first occurrence of the button component, subsequent ones will have no backtick

michaldudak reviewed (approved):
It looks good! Just one small remark from me.

danilo-leal commented on docs/data/base/getting-started/quickstart/quickstart.md:
```diff
@@ -0,0 +1,235 @@
+## Implementing a Button
+
+This is a quick tutorial that goes through the basics of using and styling Base UI components by replicating a button from GitHub's UI, using their [Primer design system](https://primer.style) as a reference.
+Base UI components help you abstract away low-level UI implementation details such as accessibility patterns, cross-browser compatibility, event handling, and more, while remaining unopinionated about styling, giving you complete control over your app's CSS.
```
I'd be down for placing this paragraph somewhere else, folks might be convinced of using Base UI already if they're diving into the Quickstart doc. But if we were to keep it, also down for your title suggestions there 🤙

mj12albert replied:
Let me take it out for now, and we can add it separately to the Overview page as Sam previously suggested !
